File: go-internal/identity.go
package main

import (
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
	"time"
)

// IssueToken performs Step 2: Token Issue (Basic Auth + GCM).
func IssueToken(config *Config, httpClient *http.Client, session *SessionContext, plaintext string) StepResult {
	basicAuth := fmt.Sprintf("Basic %s",
		base64.StdEncoding.EncodeToString([]byte(fmt.Sprintf("%s:%s", config.ClientID, config.ClientSecret))))
	auth := map[string]string{"Authorization": basicAuth}

	return postEncrypted(config, httpClient, "/v1/token/issue", session, plaintext, auth, 2, "Token Issue (Basic Auth + GCM)")
}

// IntrospectToken performs Step 3: Token Introspection (Bearer + GCM).
func IntrospectToken(config *Config, httpClient *http.Client, session *SessionContext, plaintext, accessToken string) StepResult {
	auth := map[string]string{"Authorization": fmt.Sprintf("Bearer %s", accessToken)}
	return postEncrypted(config, httpClient, "/v1/introspect", session, plaintext, auth, 3, "Token Introspection (Bearer + GCM)")
}

// RefreshToken performs Step 5: Token Refresh (Bearer + GCM).
func RefreshToken(config *Config, httpClient *http.Client, session *SessionContext, plaintext, accessToken string) StepResult {
	auth := map[string]string{"Authorization": fmt.Sprintf("Bearer %s", accessToken)}
	return postEncrypted(config, httpClient, "/v1/token", session, plaintext, auth, 5, "Token Refresh (Bearer + GCM)")
}

// RevokeToken performs Step 6: Token Revocation (Bearer + GCM).
func RevokeToken(config *Config, httpClient *http.Client, session *SessionContext, plaintext, accessToken string) StepResult {
	auth := map[string]string{"Authorization": fmt.Sprintf("Bearer %s", accessToken)}
	return postEncrypted(config, httpClient, "/v1/revoke", session, plaintext, auth, 6, "Token Revocation (Bearer + GCM)")
}

// postEncrypted: core encrypted POST — encrypt request, send, decrypt response.
func postEncrypted(config *Config, httpClient *http.Client, path string, session *SessionContext, plaintext string, authHeaders map[string]string, stepNum int, stepName string) StepResult {
	start := time.Now()

	nonce := generateNonce()
	ts := millisNow()
	aad := buildAAD(ts, nonce, session.Kid, session.ClientID)

	encrypted, err := gcmEncrypt(plaintext, session.SessionKey, aad)
	if err != nil {
		return errorResult(stepNum, stepName, plaintext, start, err.Error())
	}

	headers := map[string]string{
		"Content-Type":      "application/json",
		"X-Kid":             session.Kid,
		"X-Idempotency-Key": fmt.Sprintf("%s.%s", ts, nonce),
		"X-Clientid":        session.ClientID,
	}
	for k, v := range authHeaders {
		headers[k] = v
	}

	requestBody, _ := json.Marshal(EncryptedPayload{Payload: encrypted})

	url := fmt.Sprintf("%s/api%s", config.SidecarURL, path)
	req, err := http.NewRequest("POST", url, strings.NewReader(string(requestBody)))
	if err != nil {
		return errorResult(stepNum, stepName, plaintext, start, err.Error())
	}
	for k, v := range headers {
		req.Header.Set(k, v)
	}

	resp, err := httpClient.Do(req)
	if err != nil {
		return errorResult(stepNum, stepName, plaintext, start, err.Error())
	}
	defer resp.Body.Close()

	respHeaders := extractResponseHeaders(resp)
	respBodyBytes, _ := io.ReadAll(resp.Body)
	responseBodyStr := string(respBodyBytes)

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return StepResult{
			Step:                  stepNum,
			Name:                  stepName,
			RequestHeaders:        headers,
			RequestBodyPlaintext:  plaintext,
			RequestBodyEncrypted:  encrypted,
			ResponseHeaders:       respHeaders,
			ResponseBodyEncrypted: "",
			ResponseBodyDecrypted: responseBodyStr,
			DurationMs:            time.Since(start).Milliseconds(),
			Success:               false,
			Error:                 strPtr(fmt.Sprintf("HTTP %d: %s", resp.StatusCode, responseBodyStr)),
		}
	}

	respEncrypted, decrypted := decryptResponse(responseBodyStr, respHeaders, session)

	return StepResult{
		Step:                  stepNum,
		Name:                  stepName,
		RequestHeaders:        headers,
		RequestBodyPlaintext:  plaintext,
		RequestBodyEncrypted:  encrypted,
		ResponseHeaders:       respHeaders,
		ResponseBodyEncrypted: respEncrypted,
		ResponseBodyDecrypted: decrypted,
		DurationMs:            time.Since(start).Milliseconds(),
		Success:               true,
		Error:                 nil,
	}
}

func decryptResponse(responseBody string, respHeaders map[string]string, session *SessionContext) (string, string) {
	respKid, hasKid := respHeaders["x-kid"]
	respIdemp, hasIdemp := respHeaders["x-idempotency-key"]

	if hasKid && hasIdemp {
		parts := strings.SplitN(respIdemp, ".", 2)
		if len(parts) == 2 {
			respAAD := buildAAD(parts[0], parts[1], respKid, session.ClientID)
			var payload EncryptedPayload
			if err := json.Unmarshal([]byte(responseBody), &payload); err == nil {
				if decrypted, err := gcmDecrypt(payload.Payload, session.SessionKey, respAAD); err == nil {
					return payload.Payload, decrypted
				}
			}
		}
	}
	return "", responseBody
}

func extractResponseHeaders(resp *http.Response) map[string]string {
	headers := make(map[string]string)
	for _, key := range []string{"x-kid", "x-idempotency-key", "content-type"} {
		if val := resp.Header.Get(key); val != "" {
			headers[key] = val
		}
	}
	return headers
}

func errorResult(stepNum int, stepName, plaintext string, start time.Time, errMsg string) StepResult {
	return StepResult{
		Step:                  stepNum,
		Name:                  stepName,
		RequestHeaders:        map[string]string{},
		RequestBodyPlaintext:  plaintext,
		RequestBodyEncrypted:  "",
		ResponseHeaders:       map[string]string{},
		ResponseBodyEncrypted: "",
		ResponseBodyDecrypted: "",
		DurationMs:            time.Since(start).Milliseconds(),
		Success:               false,
		Error:                 strPtr(errMsg),
	}
}

func strPtr(s string) *string {
	return &s
}

File: rust-internal/src/identity.rs
use std::collections::HashMap;
use std::time::{Instant, SystemTime};

use base64::engine::general_purpose::STANDARD as B64;
use base64::Engine;

use crate::config::Config;
use crate::crypto;
use crate::models::{EncryptedPayload, SessionContext, StepResult};

/// Step 2: Token Issue (Basic Auth + GCM).
pub async fn issue_token(
    config: &Config,
    http_client: &reqwest::Client,
    session: &SessionContext,
    plaintext: &str,
) -> StepResult {
    let basic_auth = format!(
        "Basic {}",
        B64.encode(format!("{}:{}", config.client_id, config.client_secret))
    );
    let mut auth = HashMap::new();
    auth.insert("Authorization".to_string(), basic_auth);

    post_encrypted(
        config,
        http_client,
        "/v1/token/issue",
        session,
        plaintext,
        auth,
        2,
        "Token Issue (Basic Auth + GCM)",
    )
    .await
}

/// Step 3: Token Introspection (Bearer + GCM).
pub async fn introspect_token(
    config: &Config,
    http_client: &reqwest::Client,
    session: &SessionContext,
    plaintext: &str,
    access_token: &str,
) -> StepResult {
    let mut auth = HashMap::new();
    auth.insert(
        "Authorization".to_string(),
        format!("Bearer {access_token}"),
    );

    post_encrypted(
        config,
        http_client,
        "/v1/introspect",
        session,
        plaintext,
        auth,
        3,
        "Token Introspection (Bearer + GCM)",
    )
    .await
}

/// Step 5: Token Refresh (Bearer + GCM).
pub async fn refresh_token(
    config: &Config,
    http_client: &reqwest::Client,
    session: &SessionContext,
    plaintext: &str,
    access_token: &str,
) -> StepResult {
    let mut auth = HashMap::new();
    auth.insert(
        "Authorization".to_string(),
        format!("Bearer {access_token}"),
    );

    post_encrypted(
        config,
        http_client,
        "/v1/token",
        session,
        plaintext,
        auth,
        5,
        "Token Refresh (Bearer + GCM)",
    )
    .await
}

/// Step 6: Token Revocation (Bearer + GCM).
pub async fn revoke_token(
    config: &Config,
    http_client: &reqwest::Client,
    session: &SessionContext,
    plaintext: &str,
    access_token: &str,
) -> StepResult {
    let mut auth = HashMap::new();
    auth.insert(
        "Authorization".to_string(),
        format!("Bearer {access_token}"),
    );

    post_encrypted(
        config,
        http_client,
        "/v1/revoke",
        session,
        plaintext,
        auth,
        6,
        "Token Revocation (Bearer + GCM)",
    )
    .await
}

/// Core encrypted POST: encrypt request, send, decrypt response.
async fn post_encrypted(
    config: &Config,
    http_client: &reqwest::Client,
    path: &str,
    session: &SessionContext,
    plaintext: &str,
    auth_headers: HashMap<String, String>,
    step_num: u8,
    step_name: &str,
) -> StepResult {
    let start = Instant::now();

    let nonce = crypto::generate_nonce();
    let ts = millis_now();
    let aad = crypto::build_aad(&ts, &nonce, &session.kid, &session.client_id);

    // Encrypt request body
    let encrypted = match crypto::encrypt(plaintext, &session.session_key, &aad) {
        Ok(enc) => enc,
        Err(e) => return error_result(step_num, step_name, plaintext, &start, &e),
    };

    // Build request headers
    let mut headers = HashMap::new();
    headers.insert("Content-Type".to_string(), "application/json".to_string());
    headers.insert("X-Kid".to_string(), session.kid.clone());
    headers.insert("X-Idempotency-Key".to_string(), format!("{ts}.{nonce}"));
    headers.insert("X-ClientId".to_string(), session.client_id.clone());
    for (k, v) in &auth_headers {
        headers.insert(k.clone(), v.clone());
    }

    let request_body = serde_json::to_string(&EncryptedPayload {
        payload: encrypted.clone(),
    })
    .unwrap();

    // Send HTTP POST
    let url = format!("{}/api{path}", config.sidecar_url);
    let mut req = http_client.post(&url);
    for (k, v) in &headers {
        if k.eq_ignore_ascii_case("Content-Type") {
            continue;
        }
        req = req.header(k.as_str(), v.as_str());
    }
    req = req
        .header("Content-Type", "application/json")
        .body(request_body);

    let response = match req.send().await {
        Ok(r) => r,
        Err(e) => {
            return error_result(step_num, step_name, plaintext, &start, &e.to_string());
        }
    };

    // Extract response headers
    let resp_headers = extract_response_headers(&response);
    let status = response.status();
    let response_body_str = response.text().await.unwrap_or_default();

    if !status.is_success() {
        return StepResult {
            step: step_num,
            name: step_name.to_string(),
            request_headers: headers,
            request_body_plaintext: plaintext.to_string(),
            request_body_encrypted: encrypted,
            response_headers: resp_headers,
            response_body_encrypted: String::new(),
            response_body_decrypted: response_body_str.clone(),
            duration_ms: start.elapsed().as_millis(),
            success: false,
            error: Some(format!("HTTP {status}: {response_body_str}")),
        };
    }

    // Decrypt response
    let (resp_encrypted, decrypted) =
        decrypt_response(&response_body_str, &resp_headers, session);

    StepResult {
        step: step_num,
        name: step_name.to_string(),
        request_headers: headers,
        request_body_plaintext: plaintext.to_string(),
        request_body_encrypted: encrypted,
        response_headers: resp_headers,
        response_body_encrypted: resp_encrypted,
        response_body_decrypted: decrypted,
        duration_ms: start.elapsed().as_millis(),
        success: true,
        error: None,
    }
}

fn decrypt_response(
    response_body: &str,
    resp_headers: &HashMap<String, String>,
    session: &SessionContext,
) -> (String, String) {
    let resp_kid = resp_headers.get("x-kid");
    let resp_idemp = resp_headers.get("x-idempotency-key");

    if let (Some(kid), Some(idemp)) = (resp_kid, resp_idemp) {
        let parts: Vec<&str> = idemp.splitn(2, '.').collect();
        if parts.len() == 2 {
            let resp_aad = crypto::build_aad(parts[0], parts[1], kid, &session.client_id);

            if let Ok(payload) = serde_json::from_str::<EncryptedPayload>(response_body) {
                if let Ok(decrypted) =
                    crypto::decrypt(&payload.payload, &session.session_key, &resp_aad)
                {
                    return (payload.payload, decrypted);
                }
            }
        }
    }

    (String::new(), response_body.to_string())
}

fn extract_response_headers(response: &reqwest::Response) -> HashMap<String, String> {
    let mut headers = HashMap::new();
    for key in ["x-kid", "x-idempotency-key", "content-type"] {
        if let Some(val) = response.headers().get(key) {
            if let Ok(s) = val.to_str() {
                headers.insert(key.to_string(), s.to_string());
            }
        }
    }
    headers
}

fn error_result(
    step_num: u8,
    step_name: &str,
    plaintext: &str,
    start: &Instant,
    error: &str,
) -> StepResult {
    StepResult {
        step: step_num,
        name: step_name.to_string(),
        request_headers: HashMap::new(),
        request_body_plaintext: plaintext.to_string(),
        request_body_encrypted: String::new(),
        response_headers: HashMap::new(),
        response_body_encrypted: String::new(),
        response_body_decrypted: String::new(),
        duration_ms: start.elapsed().as_millis(),
        success: false,
        error: Some(error.to_string()),
    }
}

fn millis_now() -> String {
    SystemTime::now()
        .duration_since(SystemTime::UNIX_EPOCH)
        .unwrap()
        .as_millis()
        .to_string()
}
